taker-frontend/src/components/CloseButton.tsx

```diff
@@ -0,0 +1,129 @@
+import {
+    Box,
+    Button,
+    Popover,
+    PopoverArrow,
+    PopoverBody,
+    PopoverCloseButton,
+    PopoverContent,
+    PopoverFooter,
+    PopoverHeader,
+    PopoverTrigger,
+    Text,
+} from "@chakra-ui/react";
+import * as React from "react";
+import { Cfd, StateGroupKey, StateKey } from "../types";
+import usePostRequest from "../usePostRequest";
+
+interface CloseButtonProps {
+    cfd: Cfd;
+    connectedToMaker: Boolean;
+}
+
+/// Button that can will trigger non-collaborative close if the maker is offline and collaborative close if the maker is online
+export const CloseButton = ({ cfd, connectedToMaker }: CloseButtonProps) => {
+    let [settle, isSettling] = usePostRequest(`/api/cfd/${cfd.order_id}/settle`);
+    let [commit, isCommiting] = usePostRequest(`/api/cfd/${cfd.order_id}/commit`);
+
+    const disableCloseButton = cfd.state.getGroup() === StateGroupKey.CLOSED
+        || !(cfd.state.key === StateKey.OPEN);
+
+    let closeButton;
+
+    if (connectedToMaker) {
+        closeButton = <Box w={"45%"}>
```

Comment:
it's really hard to reason about code and maintain it when there's a lot of duplication inside it, with small changes (ie. which action gets triggered, the button name, and the text). IMHO it would be much nicer if you extracted the duplicated elements into a component, instead of a component that hides the duplication.

consider something like this:

```ts
if (connectedToMaker) {
   <CloseButton action={settle} [...] />
} 
   else {
  <CloseButton action={commit} [...] />
}

and then you wouldn't need to move the post requests inside the new component, as they're just passed in as props.

Reply:
Agreed! Bonus-points for immutability:

```tsx
const closeButton = connectedToMaker ? <CloseButton action={settle} [...] /> : <CloseButton action={commit} [...] />;
```

Reply:
I think I will still have to pass in the `connectedToMaker` bool to decide which text is shown.  don't think it is a good idea to pass in `connectedToMaker` bool and the post request as it implies that you can settle when `connectedToMaker` is false.

I guess I could also pass this through but it might be a bit too much.